File: scala/Loops.scala
object Loops {
  def main(args: Array[String]) = {
    // constants
    val start = System.currentTimeMillis()
    val iterations = 500000
    val sampleSize = 23

    // create data
    val random = scala.util.Random
    val createDataPoint = (_: Int) => random.nextInt(365)
    val createSample = (_: Int) => (1 to sampleSize).map(createDataPoint)
    val data = (1 to iterations).map(createSample)

    // calcs
    val count = data.filter(sample => (sample.distinct.length != sampleSize)).length
    val percent = count.toFloat / iterations * 100
    val finish = System.currentTimeMillis()
    val seconds = (finish - start).toFloat / 1000

    // output
    println(s"iterations: $iterations")
    println(s"sample-size: $sampleSize")
    println(f"percent: $percent%.2f")
    println(f"seconds: $seconds%.3f")
  }
}

File: lua/loops.lua
-- constants
start = os.clock()
iterations = 200000
sample_size = 23
count = 0

-- generate data
for i=1,iterations do
  data = {}
  for s=1,sample_size do
    sample = math.random(0, 365)
    if data[sample] == 1 then
      count = count + 1
      break
    else
      data[sample] = 1
    end
  end
end

-- calcs
percent = count / iterations * 100
finish = os.clock()
seconds = finish - start

-- output
print(string.format("iterations: %d", iterations))
print(string.format("sample-size: %d", sample_size))
print(string.format("percent: %.2f", percent))
print(string.format("seconds: %.3f", seconds))
